mllam_data_prep/derived_variables.py

```diff
@@ -0,0 +1,459 @@
+import importlib
+import sys
+
+import numpy as np
+import xarray as xr
+from loguru import logger
+
+
+def derive_variables(fp, derived_variables, chunking):
+    """
+    Load the dataset, and derive the specified variables
+
+    Parameters
+    ---------
+    fp : str
+        Filepath to the source dataset, for example the path to a zarr dataset
+        or a netCDF file (anything that is supported by `xarray.open_dataset` will work)
+    derived_variables : dict
+        Dictionary with the variables to derive
+        with keys as the variable names and values with entries for
+        kwargs and function to use in the calculation
+    chunking: dict
+        Dictionary with keys as the dimensions to chunk along and values
+        with the chunk size
+
+    Returns
+    -------
+    ds : xr.Dataset
+        Dataset with derived variables included
+    """
+    logger.info("Deriving variables")
+
+    try:
+        ds = xr.open_zarr(fp)
+    except ValueError:
+        ds = xr.open_dataset(fp)
+
+    ds_subset = xr.Dataset()
+    ds_subset.attrs.update(ds.attrs)
+    for _, derived_variable in derived_variables.items():
+        required_kwargs = derived_variable.kwargs
+        function_name = derived_variable.function
+        derived_variable_attributes = derived_variable.attributes or {}
+
+        # Separate the lat,lon from the required variables as these will be derived separately
+        latlon_coords_to_include = {}
+        for k, v in list(required_kwargs.items()):
+            if k in ["lat", "lon"]:
+                latlon_coords_to_include[k] = required_kwargs.pop(k)
+
+        # Subset the dataset
+        ds_input = ds[required_kwargs.keys()]
+
+        # Any coordinates needed for the derivation, for which chunking should be performed,
+        # should be converted to variables since it is not possible for *indexed* coordinates
+        # to be chunked dask arrays
+        chunks = {d: chunking.get(d, int(ds_input[d].count())) for d in ds_input.dims}
+        required_coordinates = [
+            req_var for req_var in required_kwargs.keys() if req_var in ds_input.coords
+        ]
+        ds_input = ds_input.drop_indexes(required_coordinates, errors="ignore")
+        for req_coord in required_coordinates:
+            if req_coord in chunks:
+                ds_input = ds_input.reset_coords(req_coord)
+
+        # Chunk the dataset
+        ds_input = _chunk_dataset(ds_input, chunks)
+
+        # Add function arguments to kwargs
+        kwargs = {}
+        if len(latlon_coords_to_include):
+            latlon = get_latlon_coords_for_input(ds)
```

Comment:
The dummy function is preparation for #38, in which it will be possible to derive the `lat` and `lon` coordinates from projection information if they are not included in the dataset.